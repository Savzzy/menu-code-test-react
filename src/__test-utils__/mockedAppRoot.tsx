import { render, RenderResult } from "@testing-library/react";
import React from "react";
import { Provider } from "react-redux";
import { createStore } from "redux";
import { ThemeProvider } from "styled-components";
import globalTheme from "../globalTheme";
import rootReducer from "../reducers";
import { intialStore } from "../store";
import { Store } from "../types";

interface MockedAppRootOptions {
  mockStore: Store;
}

const mockedAppRoot = (
  ComponentTree: JSX.Element,
  options?: MockedAppRootOptions,
): RenderResult => {
  const AppContainer: React.FC = ({ children }) => (
    <Provider
      store={createStore(rootReducer, options?.mockStore || intialStore)}
    >
      <ThemeProvider theme={globalTheme}>{children}</ThemeProvider>
    </Provider>
  );

  return render(ComponentTree, { wrapper: AppContainer });
};

export default mockedAppRoot;
