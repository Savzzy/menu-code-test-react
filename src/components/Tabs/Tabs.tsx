import React from "react";
import styled from "styled-components";

export enum TabPositions {
  center = "center",
  flexStart = "flex-start",
  flexEnd = "flex-end",
}

interface TabContainerProps {
  justifyContent?: TabPositions;
}

const TabContainer = styled.div<TabContainerProps>`
  width: 100%;
  margin-right: 15px;
  display: flex;
  border-bottom-color: ${(props) => props.theme.colors.seperator};
  border-bottom-width: 1px;
  border-bottom-style: solid;
  justify-content: ${(props) => props.justifyContent || "flex-start"};
`;

interface TabProps {
  selected?: boolean;
}

const Tab = styled.div<TabProps>`
  text-align: center;
  padding: 20px;
  font-size: 20px;
  text-transform: capitalize;
  cursor: pointer;
  border-bottom-color: ${(props) =>
    props.selected ? props.theme.colors.primary : props.theme.colors.white};
  border-bottom-width: 4px;
  border-bottom-style: solid;
  transition: border-bottom-color 0.25s;
`;

interface TabsProps extends TabContainerProps {
  tabOptions: Array<string>;
  selected: string;
  onTabClick: (tabOption: string) => void;
}

const Tabs: React.FC<TabsProps> = ({
  tabOptions,
  selected,
  onTabClick,
  justifyContent,
}) => {
  return (
    <TabContainer justifyContent={justifyContent}>
      {tabOptions.map((tabOption: string) => {
        return (
          <Tab
            key={tabOption}
            selected={tabOption === selected}
            onClick={() => {
              onTabClick(tabOption);
            }}
            data-testid={tabOption}
          >
            {tabOption}
          </Tab>
        );
      })}
    </TabContainer>
  );
};

export default Tabs;
