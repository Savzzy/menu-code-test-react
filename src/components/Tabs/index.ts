import Tabs from "./Tabs";
export { TabPositions } from "./Tabs";

export default Tabs;
