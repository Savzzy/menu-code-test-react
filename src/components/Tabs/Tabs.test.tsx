import React from "react";
import mockedAppRoot from "../../__test-utils__/mockedAppRoot";
import Tabs from "./Tabs";
import { fireEvent } from "@testing-library/react";

describe("Tabs", () => {
  it("renders Tab component correctly", () => {
    const myMock = jest.fn();
    const { container } = mockedAppRoot(
      <Tabs
        tabOptions={["mains", "desserts", "starters"]}
        selected={"mains"}
        onTabClick={myMock}
      />,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it("functions properly when selected Tab Clicked", () => {
    const myMock = jest.fn();
    const { getByTestId } = mockedAppRoot(
      <Tabs
        tabOptions={["mains", "desserts", "starters"]}
        selected={"mains"}
        onTabClick={myMock}
      />,
    );
    fireEvent.click(getByTestId("mains"));
    expect(myMock.mock.calls.length).toBe(1);
    expect(myMock.mock.calls[0][0]).toBe("mains");
  });
});
