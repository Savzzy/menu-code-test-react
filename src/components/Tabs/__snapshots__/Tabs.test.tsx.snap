// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs renders Tab component correctly 1`] = `
<div
  class="sc-AxjAm dWrqPI"
>
  <div
    class="sc-AxirZ hZPaiT"
    data-testid="mains"
  >
    mains
  </div>
  <div
    class="sc-AxirZ jwJYtR"
    data-testid="desserts"
  >
    desserts
  </div>
  <div
    class="sc-AxirZ jwJYtR"
    data-testid="starters"
  >
    starters
  </div>
</div>
`;
